### PR TITLE
Name collision refactor

### LINKED ISSUE #488: `Self` in paths can only be used in start position
When implementing the Cloudformation model, i get this error during the codegen gradle job

```

    Checking aws-sdk-cloudformation v0.0.7-alpha (/Users/mclean/code/smithy-rs/aws/sdk/build/aws-sdk/cloudformation)
error: expected identifier, found keyword `Self`
   --> cloudformation/src/model.rs:142:5
    |
142 |     Self,
    |     ^^^^ expected identifier, found keyword

error[E0433]: failed to resolve: `Self` in paths can only be used in start position
   --> cloudformation/src/model.rs:142:5
    |
142 |     Self,
    |     ^^^^ `Self` in paths can only be used in start position

error[E0433]: failed to resolve: `Self` in paths can only be used in start position
   --> cloudformation/src/model.rs:139:27
    |
139 | #[non_exhaustive]#[derive(std::clone::Clone, std::cmp::Eq, std::cmp::Ord, std::cmp::PartialEq, std::cmp::PartialOrd, std::fmt::Debug, std...
    |                           ^^^^^^^^^^^^^^^^^ `Self` in paths can only be used in start position

error[E0433]: failed to resolve: `Self` in paths can only be used in start position
   --> cloudformation/src/model.rs:149:31
    |
149 |             "SELF" => CallAs::Self,
    |                               ^^^^ `Self` in paths can only be used in start position

error[E0433]: failed to resolve: `Self` in paths can only be used in start position
   --> cloudformation/src/model.rs:165:21
    |
165 |             CallAs::Self => "SELF",
    |                     ^^^^ `Self` in paths can only be used in start position

error: aborting due to 5 previous errors

```

### PR DESCRIPTION
**Pull request by rcoh**
*Issue #, if available:* Fixes #488 
*Description of changes:* A medium sized overhaul of our renaming machinery.

1. Renaming was centralized to `ReservedWordSymbolProvider`
2. `Unknown` was un-special cased and docs are now generated generically during renames
3. Support the case where a service defines `UnknownValue`
4. Fix integration test which wasn't actually generating this enum.

Generated code from integration test:
```rust
pub enum UnknownVariantCollidingEnum {
    Known,
    /// **NOTE:** `::Self` has been renamed to `::SelfValue`.
    SelfValue,
    /// **NOTE:** `::SelfValue` has been renamed to `::SelfValue_`.
    SelfValue_,
    /// **NOTE:** `::Unknown` has been renamed to `::UnknownValue`.
    UnknownValue,
    /// **NOTE:** `::UnknownValue` has been renamed to `::UnknownValue_`.
    UnknownValue_,
    /// Unknown contains new variants that have been added since this code was generated.
    Unknown(String),
}
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/RustReservedWords.kt [hunk 1]
```diff
@@ -22,6 +25,25 @@ class RustReservedWordSymbolProvider(base: RustSymbolProvider) : WrappingSymbolP
     override fun toSymbol(shape: Shape): Symbol {
         return internal.toSymbol(shape)
     }
+
+    override fun toEnumVariantName(definition: EnumDefinition): MaybeRenamed? {
+        val baseName = base.toEnumVariantName(definition) ?: return null
+        if (baseName.renamedFrom != null) {
+            throw CodegenException("hmm")
```
Comment (jdisanti) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/rustlang/RustReservedWords.kt:32]: What's the intent here?

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/EnumGenerator.kt [hunk 2]
```diff
@@ -142,20 +131,20 @@ class EnumGenerator(
...
+            sortedMembers.mapNotNull { it.name() }.filter { it.renamedFrom != null }.joinToString("\n") {
+                val previousName = it.renamedFrom!!
+                "`$enumName::$previousName` has been renamed to `::${it.name}`."
+            }
         writer.docWithNote(
             shape.getTrait<DocumentationTrait>()?.value,
-            when (sortedMembers.any { it.collidesWithUnknown }) {
-                true ->
-                    "`$enumName::$UnknownVariant` has been renamed to `::$EscapedUnknownVariant`. " +
-                        "`$enumName::$UnknownVariant` refers to additional values that may have been added since " +
```
Comment (jdisanti) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/EnumGenerator.kt:150]: Does the `Unknown` variant have documentation on it indicating what it is?

Comment (rcoh) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/EnumGenerator.kt:150]: yeah added (see description)

**Review by jdisanti — APPROVED** [review]
LGTM!

**Pull request merged**